In [ ]:
from google.colab import drive
import os

drive_path = '/content/drive/'

drive.mount(drive_path, force_remount=False)

In [ ]:
working_path = os.path.join(drive_path, 'My Drive/이어드림2/')

if os.path.exists(working_path):
    print(os.listdir(working_path))
else:
    print(working_path, '경로 없음.')

In [ ]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

import warnings

warnings.filterwarnings('ignore')
plt.rcParams['font.family'] = 'Malgun Gothic'
plt.rcParams['axes.unicode_minus'] = False

# 데이터 가져오기

In [ ]:
df_cols_info = pd.read_excel(os.path.join(working_path, '기업 CB 데이터 항목설명.xlsx'))
df_cols_info

,데이터셋명,Unnamed: 1,Unnamed: 2,Unnamed: 3
0,컬럼명,형식,값 예시,설명
1,기준년월,문자(length : 8),20210901,NaN
2,가명식별자,문자(length : 20),2546,비식별화코드
3,업종(중분류),문자(length : 3),N75,업종(중분류) 코드 유효값 정의
4,외감구분,문자(length : 1),1,"1: 외감, 2:비외감"
...,...,...,...,...
156,"공공정보(국세,지방세,관세체납,고용산재체납)건수(CIS)(5년내발생)",숫자(length : 12),1,NaN
157,"신용도판단정보공공정보최근발생일자로부터경과일수(CIS)(해제,삭제)",숫자(length : 12),999999999,NaN
158,"신용도판단정보공공정보최근해제일자로부터경과일수(CIS)(해제,삭제)",숫자(length : 12),999999999,NaN
159,기업신용평가등급(구간화),문자(length : 2),4,10구간으로 구분(1~10. 1등급일수록 우량)


In [ ]:
df = pd.read_csv(os.path.join(working_path, '기업신용평가정보_합성데이터.csv'), encoding='cp949')
df

,기준년월,업종(중분류),외감구분,설립일자,종업원수,주소지시군구,상장일자,상장폐지일자,유동자산,비유동자산,...,신용도판단정보공공정보건수(CIS)(미해제),신용도판단정보공공정보건수(관련인제외)(CIS)(당월유지)(해제포함),"공공정보(국세,지방세,관세체납)건수(CIS)(미해제)","공공정보(국세,지방세,관세체납)건수(CIS)(5년내발생)","공공정보(국세,지방세,관세체납,고용산재체납)건수(CIS)(미해제)","공공정보(국세,지방세,관세체납,고용산재체납)건수(CIS)(5년내발생)","신용도판단정보공공정보최근발생일자로부터경과일수(CIS)(해제,삭제)","신용도판단정보공공정보최근해제일자로부터경과일수(CIS)(해제,삭제)",기업신용평가등급(구간화),모형개발용Performance(향후1년내부도여부)
0,20220701,C26,1,19690818,116097,41117.0,19750829.0,NaN,1.504270e+11,3.596010e+11,...,0,0,0,0,0,0,999999999,999999999,2,0
1,20220601,C26,1,19690713,121404,41117.0,19741123.0,NaN,1.526740e+11,3.566620e+11,...,0,0,0,0,0,0,999999999,999999999,2,0
2,20220501,C26,1,19690107,125782,41117.0,19741011.0,NaN,1.490870e+11,3.552700e+11,...,0,0,0,0,0,0,999999999,999999999,2,0
3,20220401,C26,1,19690823,116120,41117.0,19741213.0,NaN,1.471070e+11,3.562730e+11,...,0,0,0,0,0,0,999999999,999999999,2,0
4,20220301,C26,1,19681122,121404,41117.0,19750112.0,NaN,1.448420e+11,3.553570e+11,...,0,0,0,0,0,0,999999999,999999999,2,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
599995,20220701,H52,1,20180105,15,NaN,NaN,NaN,2.148230e+07,-1.650621e+08,...,0,0,0,0,0,0,999999999,999999999,4,0
599996,20220501,F41,1,20150302,98,28237.0,NaN,NaN,1.513349e+07,-1.620054e+08,...,0,0,0,0,0,0,999999999,999999999,4,0
599997,20220301,C22,1,20040707,50,11470.0,NaN,NaN,2.156300e+07,-1.818125e+08,...,0,0,0,0,0,0,999999999,999999999,1,0
599998,20210901,C26,1,19720407,12,47190.0,19970119.0,20110613.0,2.431177e+06,-1.885779e+08,...,0,0,0,0,0,0,1770,1680,7,0


# EDA

In [ ]:
for i in list(df.columns):
    print(i)

기준년월
업종(중분류)
외감구분
설립일자
종업원수
주소지시군구
상장일자
상장폐지일자
유동자산
비유동자산
당좌자산
재고자산
유형자산
재공품
현금
현금등가물
상품유가증권
현금성자산
매출채권
매출채권(전기)
매출채권처분손실(당기)
무형자산
투자자산
자산총계
자산총계(전기)
유동부채
단기차입금
차입금
매입채무
비유동부채
부채총계
자기자본(납입자본금)
자본잉여금
납입자본
이익잉여금
자본조정
기타포괄손익누계액
유보금
자본총계
전기자본총계
매출액
전기매출액
매출원가
매출총이익
판매비와관리비
법인세비용차감전순이익
전기법인세차감전순이익
법인세
계속사업이익
중단산업손익
금융비용
영업손익
전기영업이익
영업외수익
영업외비용
법인세차감전순이익
당기순이익
당기순이익(전기)
현금흐름
영업활동현금흐름
투자활동현금흐름
재무활동현금흐름
부채상환계수
영업이익이자보상배율
이자비용
사채이자(당기)
이자보상배율
적립금비율
EBIT
EBITDA
청산가치율
청산가치
순운전자본
순차입금
재무비율_총자산증가율
재무비율_부채비율
재무비율_자기자본비율
재무비율_유동비율
재무비율_차입금의존도
재무비율_매출액증가율
재무비율_영업이익율
재무비율_당기순이익율
재무비율_매출원가율
재무비율_판관비율
재무비율_자기자본이익률(ROE)
재무비율_매출채권회전율
재무비율_재고자산회전율
재무비율_매입채무회전율
재무비율_총자산회전율
재무비율_총자산순이익률
재무비율_유동자산증가율
재무비율_유형자산증가율
단기차입금의존도
당좌비율
순차입금비율
순운전자본회전율
총자본회전율
자기자본순이익율
매출총이익율
EBITDA마진율
영업이익증가율
당기순이익증가율
EBITDA증가율
OCF/매출액비용
부채상환계수.1
차입금/EBITDA
EBITDA/금융비용
사업장소유여부
소유건축물건수
소유건축물실거래가합계
사업장권리침해여부
소유건축물권리침해여부
기업신용공여연체과목수(일보)(미해제)
기업신용공여연체과목수(일보)(3개월내유지)(해제포함)
기업신용공여연체과목수(일보)(6개월내유지)(해제포함)
기업신용공여연체과목수(일보)(1년내유지)(해제포함)
기업신용공여연체과목

In [ ]:
df.기준년월.unique()

array([20220701, 20220601, 20220501, 20220401, 20220301, 20220201,
       20220101, 20211201, 20211101, 20210901, 20211001, 20210801])

`기준년월` 컬럼은 최장 1년 이내 기간폭만 가짐. 분석에 무의미하다고 판단

In [ ]:
df = df.drop(columns=['기준년월'])
# df

## 중복 행 처리

In [ ]:
print(f'Number of duplicate rows: {df.duplicated().sum()}')

Number of duplicate rows: 1537


In [ ]:
print(f'Number of duplicate rows: {df.duplicated(subset=[col for col in df.columns if col not in ['설립일자']]).sum()}')

Number of duplicate rows: 281837


`설립일자`까지도 같은 행은 1,537개, `설립일자` 외 모든 컬럼이 같은 행은 281,837개.

이것이 오기인지 여부 확인 필요!!

In [ ]:
dup_groups = df[df.duplicated(subset=[col for col in df.columns if col not in ['설립일자']], keep=False)]
dup_groups.sort_values(by=[col for col in df.columns if col not in ['설립일자']]).head()

,업종(중분류),외감구분,설립일자,종업원수,주소지시군구,상장일자,상장폐지일자,유동자산,비유동자산,당좌자산,...,신용도판단정보공공정보건수(CIS)(미해제),신용도판단정보공공정보건수(관련인제외)(CIS)(당월유지)(해제포함),"공공정보(국세,지방세,관세체납)건수(CIS)(미해제)","공공정보(국세,지방세,관세체납)건수(CIS)(5년내발생)","공공정보(국세,지방세,관세체납,고용산재체납)건수(CIS)(미해제)","공공정보(국세,지방세,관세체납,고용산재체납)건수(CIS)(5년내발생)","신용도판단정보공공정보최근발생일자로부터경과일수(CIS)(해제,삭제)","신용도판단정보공공정보최근해제일자로부터경과일수(CIS)(해제,삭제)",기업신용평가등급(구간화),모형개발용Performance(향후1년내부도여부)
210974,A01,1,20070916,1,28185.0,NaN,NaN,3181482.0,11004710.0,3162917.0,...,0,0,0,0,0,0,999999999,999999999,6,0
210975,A01,1,20080213,1,28185.0,NaN,NaN,3181482.0,11004710.0,3162917.0,...,0,0,0,0,0,0,999999999,999999999,6,0
210976,A01,1,20071120,1,28185.0,NaN,NaN,3181482.0,11004710.0,3162917.0,...,0,0,0,0,0,0,999999999,999999999,6,0
116912,A01,1,20080309,1,28185.0,NaN,NaN,4468537.0,26895549.0,2258905.0,...,0,0,0,0,0,0,640,553,6,0
116913,A01,1,20080626,1,28185.0,NaN,NaN,4468537.0,26895549.0,2258905.0,...,0,0,0,0,0,0,640,553,6,0


사실상 같은 기업이라고 판단해도 무방. 남기면 오히려 노이즈 학습 위험

In [ ]:
df = df.drop(columns=['설립일자'])\
       .drop_duplicates(keep='first')
# df

## 결측치 처리1

In [ ]:
df.isna().mean()[df.isna().sum() > 0].sort_values(ascending=False)

상장폐지일자          0.998730
상장일자            0.930900
재무비율_총자산순이익률    0.643896
재무비율_당기순이익율     0.637510
주소지시군구          0.081829
재무비율_유형자산증가율    0.077243
재무비율_유동자산증가율    0.077243
재무비율_매입채무회전율    0.077243
부채상환계수.1        0.063367
EBITDA증가율       0.040753
영업이익이자보상배율      0.000515
이자보상배율          0.000515
업종(중분류)         0.000107
dtype: float64

`상장일자`(0.93), `상장폐지일자`(0.99), `재무비율_총자산순이익률`(0.64), `재무비율_당기순이익율`(0.63) 컬럼은 절반 이상이 결측치

대체나 행 삭제 처리할 수 없음 -> 컬럼 삭제

In [ ]:
df = df.drop(columns=['상장일자', '상장폐지일자', '재무비율_총자산순이익률', '재무비율_당기순이익율'])
# df

In [ ]:
# plt.figure(figsize=(18, 6))

# sns.histplot(data=df, x='업종(중분류)')

# plt.xlabel('Biz code')
# plt.xticks(rotation=45)

`업종`, `주소지시군구` 컬럼은 기업 신용도 평가에서 중요한 데이터이므로 삭제 혹은 대체 어려움

또한 `object`, `float` 타입이지만 실제로는 `category`임

결측치 유지
1. Unknown 항목 삽입 -> 별도의 항목으로 인식
2. dtypes `category`로 변환 시 자동 결측으로 분류 -> 결측으로 인식 가능

## 데이터 타입 정리

In [ ]:
for i in range(len(df.dtypes)):
    print(df.columns[i], ":", df.dtypes[i])

업종(중분류) : object
외감구분 : int64
종업원수 : int64
주소지시군구 : float64
유동자산 : float64
비유동자산 : float64
당좌자산 : float64
재고자산 : int64
유형자산 : float64
재공품 : int64
현금 : int64
현금등가물 : int64
상품유가증권 : int64
현금성자산 : int64
매출채권 : int64
매출채권(전기) : int64
매출채권처분손실(당기) : int64
무형자산 : int64
투자자산 : float64
자산총계 : float64
자산총계(전기) : float64
유동부채 : float64
단기차입금 : int64
차입금 : int64
매입채무 : int64
비유동부채 : int64
부채총계 : float64
자기자본(납입자본금) : int64
자본잉여금 : int64
납입자본 : int64
이익잉여금 : float64
자본조정 : int64
기타포괄손익누계액 : int64
유보금 : float64
자본총계 : float64
전기자본총계 : float64
매출액 : float64
전기매출액 : float64
매출원가 : float64
매출총이익 : float64
판매비와관리비 : int64
법인세비용차감전순이익 : int64
전기법인세차감전순이익 : int64
법인세 : int64
계속사업이익 : int64
중단산업손익 : int64
금융비용 : int64
영업손익 : int64
전기영업이익 : int64
영업외수익 : int64
영업외비용 : int64
법인세차감전순이익 : int64
당기순이익 : int64
당기순이익(전기) : int64
현금흐름 : int64
영업활동현금흐름 : float64
투자활동현금흐름 : int64
재무활동현금흐름 : int64
부채상환계수 : float64
영업이익이자보상배율 : float64
이자비용 : int64
사채이자(당기) : int64
이자보상배율 : float64
적립금비율 : float64
EBIT : int64
EBITDA

In [ ]:
# df['설립일자'] = pd.to_datetime(df['설립일자'], format='%Y%m%d', errors='coerce') # 위에서 컬럼 삭제
df['업종(중분류)'] = df['업종(중분류)'].astype('category')
df['주소지시군구'] = df['주소지시군구'].astype('category')
df['외감구분'] = df['외감구분'].astype('category')

df['기업신용평가등급(구간화)'] = df['기업신용평가등급(구간화)'].astype('category')
df['모형개발용Performance(향후1년내부도여부)'] = df['모형개발용Performance(향후1년내부도여부)'].astype('category')

# print("\nData types after conversion:")
# print(df.dtypes)

## 연체 컬럼 정리

In [ ]:
for c in df.columns:
    if '연체' in c: # 새로 만드는 컬럼은 '기업신용공여' 없음
        print(c)

기업신용공여연체과목수(일보)(미해제)
기업신용공여연체과목수(일보)(3개월내유지)(해제포함)
기업신용공여연체과목수(일보)(6개월내유지)(해제포함)
기업신용공여연체과목수(일보)(1년내유지)(해제포함)
기업신용공여연체과목수(일보)(3년내유지)(해제포함)
기업신용공여연체과목수(일보)(3개월내발생)(해제포함)
기업신용공여연체과목수(일보)(6개월내발생)(해제포함)
기업신용공여연체과목수(일보)(1년내발생)(해제포함)
기업신용공여연체과목수(일보)(3년내발생)(해제포함)
기업신용공여연체과목수(일보)(3개월내유지)(연체일수30일이상)(해제포함)
기업신용공여연체과목수(일보)(6개월내유지)(연체일수30일이상)(해제포함)
기업신용공여연체과목수(일보)(1년내유지)(연체일수30일이상)(해제포함)
기업신용공여연체과목수(일보)(3년내유지)(연체일수30일이상)(해제포함)
기업신용공여30일이상연체과목수(일보)(해제포함)
기업신용공여30일이상연체과목수(일보)(미해제)
기업신용공여30일이상연체과목수(일보)(이자연체)(해제포함)
기업신용공여30일이상연체과목수(일보)(이자연체)(미해제)
기업신용공여연체기관수(일보)(미해제)
기업신용공여30일이상연체기관수(일보)(해제포함)
기업신용공여30일이상연체기관수(일보)(미해제)
기업신용공여30일이상연체기관수(일보)(연체)(해제포함)
기업신용공여30일이상연체기관수(일보)(이자연체)(미해제)
기업신용공여연체기관수(일보)(3개월내유지)(연체일수30일이상)(해제포함)
기업신용공여연체기관수(일보)(6개월내유지)(연체일수30일이상)(해제포함)
기업신용공여연체기관수(일보)(1년내유지)(연체일수30일이상)(해제포함)
기업신용공여연체기관수(일보)(3년내유지)(연체일수30일이상)(해제포함)
기업신용공여연체최장연체일수(일보)(3개월내유지)(해제포함)
기업신용공여연체최장연체일수(일보)(6개월내유지)(해제포함)
기업신용공여연체최장연체일수(일보)(1년내유지)(해제포함)
기업신용공여연체최장연체일수(일보)(3년내유지)(해제포함)
기업신용공여연체최장연체일수(일보)(5년내유지)(해제포함)
기업

연체 컬럼이 많은데, 그 중 기간별 구별에 따른 연체 기록이 많았다.

이 중 `3개월`, `6개월`, `1년`, `3년`, `5년` 중 도메인 지식의 부재로 인해 생성형 AI의 조언에 따라 `3년`을 채택했다.

또한 연체 내용 중 원금/이자 연체, 30일 미만/이상 연체로 구분되는 것을 확인했다.

이 중 이자 연체와 30일 미만 연체는 신용 평가 시 '단순 관찰'로 분류된다는 것을 생성형 AI를 통해 확인했다.

따라서 `3년 이내` 기간 중 `30일 이상` `원금 연체`의 `건수` 및 `기관수` 컬럼만을 남기기로 했다.

In [ ]:
df[df['기업신용공여30일이상연체과목수(일보)(미해제)'] < df['기업신용공여30일이상연체과목수(일보)(이자연체)(미해제)']]\
    [['기업신용공여30일이상연체과목수(일보)(미해제)', '기업신용공여30일이상연체과목수(일보)(이자연체)(미해제)']].head(4)

,기업신용공여30일이상연체과목수(일보)(미해제),기업신용공여30일이상연체과목수(일보)(이자연체)(미해제)
84,1,3
3105,1,4
20032,1,2
26406,1,2


`이자연체`가 `연체`보다 많은 것으로 보아 `연체`는 `이자연체` 미포함으로 보임

In [ ]:
# # 이상치 없는지 확인
# df[(df['총위험연체진행건수'] != 0) & (df['총위험연체진행기관수'] == 0)][['총위험연체진행건수', '총위험연체진행기관수']].shape[0], \
# df[(df['3년내기관별평균총위험연체진행건수'] != 0) & (df['3년내총위험연체기관수'] == 0)][['3년내기관별평균총위험연체진행건수', '3년내총위험연체기관수']].shape[0]

In [ ]:
df = df.rename(columns={'기업신용공여30일이상연체과목수(일보)(미해제)': '총위험연체진행건수',
                        '기업신용공여연체과목수(일보)(3년내유지)(연체일수30일이상)(해제포함)': '3년내총위험연체건수',
                        '기업신용공여30일이상연체기관수(일보)(미해제)': '총위험연체진행기관수',
                        '기업신용공여연체기관수(일보)(3년내유지)(연체일수30일이상)(해제포함)': '3년내총위험연체기관수'})

df['기관별평균총위험연체진행건수'] = np.where(df['총위험연체진행기관수'] != 0, df['총위험연체진행건수'] / df['총위험연체진행기관수'], 0)
df['3년내기관별평균총위험연체건수'] = np.where(df['3년내총위험연체기관수'] != 0, df['3년내총위험연체건수'] / df['3년내총위험연체기관수'], 0)

In [ ]:
delin_cols = [c for c in df.columns
              if '연체' in c and c not in ['총위험연체진행건수', '3년내총위험연체건수', '총위험연체진행기관수', '3년내총위험연체기관수', '기관별평균총위험연체진행건수', '3년내기관별평균총위험연체건수']]

df = df.drop(columns=delin_cols)
# print(delin_cols)

## 결측치 처리2

비율 컬럼 결측치가 남았다.

`재무비율_유형자산증가율`, `재무비율_유동자산증가율`, `재무비율_매입채무회전율`, `부채상환계수.1`, `EBITDA증가율`, `영업이익이자보상배율`, `이자보상배율`

1. 결측 비중이 크지 않으니 행 삭제 -> 결측 행이 분산된다면 비중 커짐
2. 분류 및 회귀 사용해서 대체 (KNN, MICE) -> 데이터 신뢰는 다소 하락

### 결측 행 삭제

In [ ]:
df_dropped_na = df.dropna(how='any')
df_dropped_na

,업종(중분류),외감구분,종업원수,주소지시군구,유동자산,비유동자산,당좌자산,재고자산,유형자산,재공품,...,"공공정보(국세,지방세,관세체납)건수(CIS)(미해제)","공공정보(국세,지방세,관세체납)건수(CIS)(5년내발생)","공공정보(국세,지방세,관세체납,고용산재체납)건수(CIS)(미해제)","공공정보(국세,지방세,관세체납,고용산재체납)건수(CIS)(5년내발생)","신용도판단정보공공정보최근발생일자로부터경과일수(CIS)(해제,삭제)","신용도판단정보공공정보최근해제일자로부터경과일수(CIS)(해제,삭제)",기업신용평가등급(구간화),모형개발용Performance(향후1년내부도여부),기관별평균총위험연체진행건수,3년내기관별평균총위험연체건수
0,C26,1,116097,41117.0,1.504270e+11,3.596010e+11,1.192400e+11,31186582315,2.104230e+11,0,...,0,0,0,0,999999999,999999999,2,0,0.0,0.0
1,C26,1,121404,41117.0,1.526740e+11,3.566620e+11,1.207270e+11,31946106000,2.073340e+11,0,...,0,0,0,0,999999999,999999999,2,0,0.0,0.0
2,C26,1,125782,41117.0,1.490870e+11,3.552700e+11,1.183500e+11,30737079838,2.073340e+11,0,...,0,0,0,0,999999999,999999999,2,0,0.0,0.0
3,C26,1,116120,41117.0,1.471070e+11,3.562730e+11,1.151610e+11,31946106000,2.084340e+11,0,...,0,0,0,0,999999999,999999999,2,0,0.0,0.0
4,C26,1,121404,41117.0,1.448420e+11,3.553570e+11,1.115800e+11,33262780546,2.102940e+11,0,...,0,0,0,0,999999999,999999999,2,0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
599994,F41,2,42,41173.0,1.312323e+07,-1.539277e+08,1.312323e+07,0,1.249189e+06,0,...,0,0,0,0,999999999,999999999,7,0,0.0,0.0
599996,F41,1,98,28237.0,1.513349e+07,-1.620054e+08,1.513349e+07,0,2.758494e+06,0,...,0,0,0,0,999999999,999999999,4,0,0.0,0.0
599997,C22,1,50,11470.0,2.156300e+07,-1.818125e+08,1.565215e+07,5910842,3.589975e+06,0,...,0,0,0,0,999999999,999999999,1,0,0.0,0.0
599998,C26,1,12,47190.0,2.431177e+06,-1.885779e+08,1.970343e+06,460834,2.693335e+07,0,...,0,0,0,0,1770,1680,7,0,0.0,0.0


### 결측 행 대체 (보류-시간 소요 많음)

**MICE Imputer 객체의 estimator 후보**
| 모델                        | 특징              | MICE에서의 효과                                                                  |
| ------------------------- | --------------- | --------------------------------------------------------------------------- |
| **BayesianRidge**   | 선형 회귀 + 베이지안 추정 | 선형 관계가 있는 데이터에 강함<br>안정적, 반복(iteration) 수 적어도 수렴 잘됨<br>비선형 관계 반영 어려움 |
| **RandomForestRegressor** | 앙상블 기반 비선형 회귀   | 비선형 관계를 잘 반영<br>변수 상호작용 처리 가능<br>계산 느림, 작은 데이터셋에서 과적합 가능             |
| **DecisionTreeRegressor** | 단일 결정트리         | 단순 비선형 처리 가능<br>RandomForest보다 불안정, 반복 시 변동 큼                          |
| **KNeighborsRegressor**   | 최근접 기반          | 국소적 패턴 반영<br>고차원에서는 거리 왜곡, 느림                                          |
| **LinearRegression**      | 일반 선형 회귀        | 빠르고 안정적<br>BayesianRidge보다 불확실성 처리 X                                   |
| **ExtraTreesRegressor**   | RandomForest 변형 | 비선형 관계 잘 반영, RandomForest보다 속도 빠름<br>작은 데이터에서 과적합 가능                   |

기업 재무 데이터는 단순 선형 관계보다 비선형 관계가 반영될 것으로 예상하여 RandomForest 모델 estimator 사용

In [ ]:
# from sklearn.experimental import enable_iterative_imputer
# from sklearn.impute import IterativeImputer
# from sklearn.ensemble import RandomForestRegressor

# imputable_cols = ['재무비율_유형자산증가율', '재무비율_유동자산증가율', '재무비율_매입채무회전율', '부채상환계수.1', 'EBITDA증가율', '영업이익이자보상배율', '이자보상배율']

# mice_imputer = IterativeImputer(estimator=RandomForestRegressor(n_estimators=50,
#                                                                 max_depth=10,
#                                                                 random_state=42),
#                                 max_iter=10, # 수렴까지 반복 횟수
#                                 initial_strategy='median', # 초기값: 중앙값
#                                 skip_complete=True, # 결측 없는 컬럼은 계산 스킵 (성능)
#                                 random_state=42)

MICE 대체 시 `category` 타입 컬럼 중 `기업신용평가등급(구간화)`, `모형개발용Performance(향후1년내부도여부)`는 정답 데이터이므로 제외

`업종(중분류)`, `주소지시군구`는 분류 개수가 너무 많아 폭주 가능성 있어서 제외

`외감구분` 컬럼은 0 or 1 이므로 one-hot 인코딩하여 대체 추론에 참여

In [ ]:
# df_imputed = df.copy()

# df_impute_input = pd.get_dummies(df_imputed[[c for c in df_imputed.columns if c not in ['업종(중분류)', '주소지시군구', '기업신용평가등급(구간화)', '모형개발용Performance(향후1년내부도여부)']]],
#                                  columns=['외감구분'], drop_first=False)

# df_imputed[imputable_cols] = mice_imputer.fit_transform(df_impute_input)

In [ ]:
target_cols = ['기업신용평가등급(구간화)', '모형개발용Performance(향후1년내부도여부)']

df_targets = df_dropped_na[target_cols]
df_feats = df_dropped_na[[c for c in df_dropped_na.columns if c not in target_cols]]

# 데이터 스케일링

In [ ]:
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# 범주형 및 수치형 컬럼 구분
categorical_features = ['업종(중분류)', '주소지시군구', '외감구분']
numerical_features = [col for col in df_feats.columns if col not in categorical_features]

# 전처리 파이프라인 정의
# 수치형 컬럼: StandardScaler 적용
# 범주형 컬럼: OneHotEncoder 적용 (unknown_value='ignore'를 사용하여 학습 데이터에 없는 범주 처리)
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numerical_features),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_features)
    ])

# 전처리 적용
df_processed_feats = preprocessor.fit_transform(df_feats)

# 처리된 피처의 형태 확인 (희소 행렬일 수 있음)
print("Shape of processed features:", df_processed_feats.shape)

# 모델 구축